caddyhttp/httpserver/replacer.go

```diff
@@ -132,6 +148,31 @@ func NewReplacer(r *http.Request, rr *ResponseRecorder, emptyValue string) Repla
 	return rep
 }
 
+func canLogRequest(r *http.Request) (canLog bool) {
+	if r.Method == "POST" || r.Method == "PUT" {
+		for _, cType := range r.Header[headerContentType] {
+			// the cType could have charset and other info
+			if strings.Index(cType, contentTypeJSON) > -1 || strings.Index(cType, contentTypeXML) > -1 {
+				canLog = true
+				break
+			}
+		}
+	}
+	return
+}
+
+// readRequestBody reads the request body and sets a
+// new io.ReadCloser that has not yet been read.
+func readRequestBody(r *http.Request) ([]byte, error) {
+	body, err := ioutil.ReadAll(r.Body)
```

Comment:
This should probably be an io.LimitReader, like 1 MB or even just 100 KB. Not sure what would be ideal yet.